# Notebook 05 — PostgreSQL Rate History Store

Explain why SQL is needed alongside vector search:

| Question Type              | Vector Search | SQL Query |
|----------------------------|---------------|-----------|
| "Find similar items"       | ✅ Perfect    | ❌ Can't  |
| "Avg rate for brick work"  | ❌ Can't      | ✅ Perfect |
| "Rate range across projects"| ❌ Can't     | ✅ Perfect |
| "How many times quoted?"   | ❌ Can't      | ✅ Perfect |
| "Cheapest source file?"    | ❌ Can't      | ✅ Perfect |

The agent will use BOTH:
  - BOQSearcher.smart_search() → semantic match
  - PostgreSQL queries → statistical confirmation

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path
from sqlalchemy import create_engine, text
from sqlalchemy.exc import OperationalError
import psycopg2
import json, re
from dotenv import load_dotenv
import os
load_dotenv()

CHUNKS_PATH = Path("../data/processed/boq_chunks.csv")
POSTGRES_URL = os.getenv(
    "POSTGRES_URL",
    "postgresql://rateiq:rateiq123@localhost:5432/rateiq"
)

# Connect
engine = create_engine(POSTGRES_URL)
try:
    with engine.connect() as conn:
        result = conn.execute(text("SELECT version()"))
        print(f"✓ PostgreSQL connected")
        print(f"  {result.fetchone()[0][:50]}")
except Exception as e:
    print(f"✗ Connection failed: {e}")

# Load chunks
df = pd.read_csv(CHUNKS_PATH)
print(f"\n✓ Chunks loaded: {df.shape}")
print(f"  Columns: {df.columns.tolist()}")


✓ PostgreSQL connected
  PostgreSQL 15.17 (Debian 15.17-1.pgdg13+1) on x86_

✓ Chunks loaded: (1429, 13)
  Columns: ['chunk_id', 'embedding_text', 'description_short', 'description_full', 'section_title', 'work_category', 'rate', 'unit_norm', 'qty', 'source_file', 'sheet_name', 'rate_per_unit_label', 'est_tokens']


## Step 1: Design the Schema

Explain the 2-table design:

Table 1: boq_items — one row per line item
  All 1429 items with full metadata + rate

Table 2: boq_rate_stats — pre-computed stats per item type
  Aggregated: avg/min/max rate per normalized description
  This is the "rate confirmation" table

Why 2 tables?
  boq_items   → raw data, exact lookup, audit trail
  boq_rate_stats → fast aggregation, no GROUP BY at query time

In [2]:
CREATE_ITEMS_TABLE = """
CREATE TABLE IF NOT EXISTS boq_items (
    id              SERIAL PRIMARY KEY,
    chunk_id        VARCHAR(32) UNIQUE NOT NULL,
    description_short TEXT,
    description_full  TEXT,
    section_title   TEXT,
    work_category   VARCHAR(50),
    rate            NUMERIC(15, 2),
    unit_norm       VARCHAR(20),
    qty             VARCHAR(50),
    source_file     VARCHAR(200),
    sheet_name      VARCHAR(200),
    rate_per_unit_label VARCHAR(100),
    embedding_text  TEXT,
    created_at      TIMESTAMP DEFAULT NOW()
);
"""

CREATE_STATS_TABLE = """
CREATE TABLE IF NOT EXISTS boq_rate_stats (
    id              SERIAL PRIMARY KEY,
    description_key VARCHAR(300),
    work_category   VARCHAR(50),
    unit_norm       VARCHAR(20),
    rate_count      INTEGER,
    rate_min        NUMERIC(15, 2),
    rate_max        NUMERIC(15, 2),
    rate_avg        NUMERIC(15, 2),
    rate_median     NUMERIC(15, 2),
    rate_std        NUMERIC(15, 2),
    source_files    TEXT,
    sample_rates    TEXT,
    UNIQUE(description_key, unit_norm)
);
"""

CREATE_IDX_1 = """
CREATE INDEX IF NOT EXISTS idx_boq_items_category 
ON boq_items(work_category);
"""
CREATE_IDX_2 = """
CREATE INDEX IF NOT EXISTS idx_boq_items_unit 
ON boq_items(unit_norm);
"""
CREATE_IDX_3 = """
CREATE INDEX IF NOT EXISTS idx_boq_items_rate 
ON boq_items(rate);
"""
CREATE_IDX_4 = """
CREATE INDEX IF NOT EXISTS idx_boq_stats_key 
ON boq_rate_stats(description_key);
"""

with engine.connect() as conn:
    conn.execute(text(CREATE_ITEMS_TABLE))
    conn.execute(text(CREATE_STATS_TABLE))
    conn.execute(text(CREATE_IDX_1))
    conn.execute(text(CREATE_IDX_2))
    conn.execute(text(CREATE_IDX_3))
    conn.execute(text(CREATE_IDX_4))
    conn.commit()
    print("✓ Tables created: boq_items, boq_rate_stats")
    print("✓ Indexes created: category, unit, rate, description_key")


✓ Tables created: boq_items, boq_rate_stats
✓ Indexes created: category, unit, rate, description_key


## Step 2: Insert All Items into boq_items
Using pandas to_sql for bulk insert.
ON CONFLICT DO NOTHING — safe to re-run.
chunk_id is the unique key (same as Qdrant point key).

In [3]:
# Prepare DataFrame for insert
df_insert = df.copy()

# Ensure correct column names match table schema
column_map = {
    "chunk_id":           "chunk_id",
    "description_short":  "description_short",
    "description_full":   "description_full",
    "section_title":      "section_title",
    "work_category":      "work_category",
    "rate":               "rate",
    "unit_norm":          "unit_norm",
    "qty":                "qty",
    "source_file":        "source_file",
    "sheet_name":         "sheet_name",
    "rate_per_unit_label":"rate_per_unit_label",
    "embedding_text":     "embedding_text"
}

df_insert = df_insert[[c for c in column_map 
                        if c in df_insert.columns]]
df_insert.columns = [column_map[c] for c in df_insert.columns]

# Clean rate column
df_insert["rate"] = pd.to_numeric(
    df_insert["rate"], errors="coerce"
)
df_insert = df_insert[df_insert["rate"].notna()]

print(f"Inserting {len(df_insert)} rows into boq_items...")

# Use INSERT ... ON CONFLICT DO NOTHING via raw SQL
inserted = 0
skipped  = 0

with engine.connect() as conn:
    for _, row in df_insert.iterrows():
        try:
            conn.execute(text("""
                INSERT INTO boq_items 
                    (chunk_id, description_short, description_full,
                     section_title, work_category, rate, unit_norm,
                     qty, source_file, sheet_name,
                     rate_per_unit_label, embedding_text)
                VALUES 
                    (:chunk_id, :description_short, :description_full,
                     :section_title, :work_category, :rate, :unit_norm,
                     :qty, :source_file, :sheet_name,
                     :rate_per_unit_label, :embedding_text)
                ON CONFLICT (chunk_id) DO NOTHING
            """), row.where(pd.notna(row), None).to_dict())
            inserted += 1
        except Exception as e:
            skipped += 1
    conn.commit()

print(f"✓ Inserted: {inserted}")
print(f"  Skipped (duplicates): {skipped}")

# Verify
with engine.connect() as conn:
    count = conn.execute(
        text("SELECT COUNT(*) FROM boq_items")
    ).scalar()
    print(f"  Total rows in boq_items: {count}")


Inserting 1429 rows into boq_items...
✓ Inserted: 1429
  Skipped (duplicates): 0
  Total rows in boq_items: 1346


## Step 3: Build boq_rate_stats
For each unique (description_key, unit_norm) combo,
compute: count, min, max, avg, median, std.

description_key = first 100 chars of description_short
(normalized: lowercase, strip extra whitespace)

This table answers: "How consistently is this item priced?"

In [4]:
def normalize_desc_key(text):
    """First 100 chars, lowercase, collapse whitespace."""
    if pd.isna(text):
        return "unknown"
    s = str(text).lower().strip()
    s = re.sub(r'\s+', ' ', s)
    return s[:100]

df_insert["description_key"] = df_insert[
    "description_short"
].apply(normalize_desc_key)

# Group and compute stats
stats_rows = []
grouped = df_insert.groupby(
    ["description_key", "work_category", "unit_norm"]
)

for (desc_key, category, unit), group in grouped:
    rates = group["rate"].dropna().tolist()
    if not rates:
        continue
    
    arr = np.array(rates)
    stats_rows.append({
        "description_key": desc_key,
        "work_category":   category,
        "unit_norm":       unit,
        "rate_count":      len(rates),
        "rate_min":        float(arr.min()),
        "rate_max":        float(arr.max()),
        "rate_avg":        float(arr.mean()),
        "rate_median":     float(np.median(arr)),
        "rate_std":        float(arr.std()) if len(arr) > 1 else 0.0,
        "source_files":    json.dumps(
                               group["source_file"].unique().tolist()
                           ),
        "sample_rates":    json.dumps(
                               sorted(set(rates))[:5]
                           )
    })

df_stats = pd.DataFrame(stats_rows)
print(f"Stats computed for {len(df_stats)} unique item types")
print(f"\nTop 10 most-quoted items:")
top = df_stats.nlargest(10, "rate_count")[
    ["description_key", "unit_norm", 
     "rate_count", "rate_min", "rate_max", "rate_avg"]
]
print(top.to_string(index=False))

# Insert stats
inserted_s = 0
with engine.connect() as conn:
    for _, row in df_stats.iterrows():
        conn.execute(text("""
            INSERT INTO boq_rate_stats
                (description_key, work_category, unit_norm,
                 rate_count, rate_min, rate_max, rate_avg,
                 rate_median, rate_std, source_files, sample_rates)
            VALUES
                (:description_key, :work_category, :unit_norm,
                 :rate_count, :rate_min, :rate_max, :rate_avg,
                 :rate_median, :rate_std, :source_files, :sample_rates)
            ON CONFLICT (description_key, unit_norm) DO UPDATE SET
                rate_count = EXCLUDED.rate_count,
                rate_avg   = EXCLUDED.rate_avg,
                rate_min   = EXCLUDED.rate_min,
                rate_max   = EXCLUDED.rate_max,
                rate_median = EXCLUDED.rate_median,
                rate_std   = EXCLUDED.rate_std,
                source_files = EXCLUDED.source_files
        """), row.to_dict())
        inserted_s += 1
    conn.commit()

print(f"\n✓ Stats rows inserted: {inserted_s}")


Stats computed for 1174 unique item types

Top 10 most-quoted items:
                                                                                     description_key unit_norm  rate_count      rate_min  rate_max      rate_avg
                                                                                              cement      bags          12     14.046001    1240.0    126.952021
supply and installation of veneer pressed mdf wall cladding in approved finish color and texture per       sft           8   1790.000000    1790.0   1790.000000
                                                                lifting, freight, cartage & cleaning       sft           7     10.000000      15.0     10.892857
providing, making and fixing of the wooden vanity consisting of the lamination mdf formite 3/4" for        job           5 145000.000000  380000.0 224200.000000
                                                                                           tile bond      bags           5    

## Step 4: Test SQL Queries
These are the exact queries the LangGraph agent will run.
Test all 5 agent query types.

In [5]:
def sql_rate_lookup(description_fragment: str, 
                    unit: str = None,
                    limit: int = 5) -> list[dict]:
    """
    Find items whose description contains a keyword.
    Returns rate history with source file.
    Agent uses this to find: "what rate did we use for X?"
    """
    query = """
        SELECT 
            description_short,
            rate,
            unit_norm,
            source_file,
            rate_per_unit_label,
            work_category
        FROM boq_items
        WHERE LOWER(description_short) LIKE :pattern
"""
    params = {"pattern": f"{description_fragment.lower()}%"}
    
    if unit:
        query += " AND unit_norm = :unit"
        params["unit"] = unit
    
    query += " ORDER BY rate ASC LIMIT :limit"
    params["limit"] = limit
    
    with engine.connect() as conn:
        rows = conn.execute(text(query), params).fetchall()
    
    return [dict(r._mapping) for r in rows]

# Test
print("=== SQL Query 1: Rate lookup for 'brick work' ===")
results = sql_rate_lookup("brick work", unit="sft")
for r in results:
    print(f"  {r['description_short'][:40]:40s} | "
          f"{r['rate_per_unit_label']:20s} | "
          f"{r['source_file']}")


=== SQL Query 1: Rate lookup for 'brick work' ===
  Brick Work 4.5" Thick                    | Rs. 315 per sft      | BOQ - 01 .xlsx
  Brick Work 9" Thick                      | Rs. 485 per sft      | BOQ - 01 .xlsx


In [6]:
def sql_rate_stats(description_fragment: str,
                   unit: str = None) -> dict:
    """
    Get statistical summary for an item type.
    Agent uses this for rate confirmation:
    "Brick work 4.5 inch has been quoted Rs.310-320
     across 3 projects. Average: Rs.315."
    """
    query = """
        SELECT 
            COUNT(*)                    AS count,
            MIN(rate)                   AS min_rate,
            MAX(rate)                   AS max_rate,
            ROUND(AVG(rate), 0)         AS avg_rate,
            ROUND(STDDEV(rate), 0)      AS std_rate,
            ARRAY_AGG(DISTINCT source_file) AS sources
        FROM boq_items
        WHERE LOWER(description_short) LIKE :pattern
"""
    params = {"pattern": f"{description_fragment.lower()}%"}
    if unit:
        query += " AND unit_norm = :unit"
        params["unit"] = unit
    
    with engine.connect() as conn:
        row = conn.execute(text(query), params).fetchone()
    
    if not row or row.count == 0:
        return {}
    
    return dict(row._mapping)

# Test
print("=== SQL Query 2: Stats for 'gypsum' ceiling ===")
s = sql_rate_stats("gypsum", unit="sft")
if s:
    print(f"  Count:    {s['count']} quotes")
    print(f"  Min:      Rs. {s['min_rate']:,.0f}")
    print(f"  Max:      Rs. {s['max_rate']:,.0f}")
    print(f"  Average:  Rs. {s['avg_rate']:,.0f}")
    print(f"  Std Dev:  Rs. {s['std_rate']:,.0f}")
    print(f"  Sources:  {s['sources']}")


=== SQL Query 2: Stats for 'gypsum' ceiling ===
  Count:    4 quotes
  Min:      Rs. 375
  Max:      Rs. 455
  Average:  Rs. 423
  Std Dev:  Rs. 36
  Sources:  ['3. CEILING.xlsx', 'BOQ - 01 .xlsx', 'BOQ - 05 .xlsx']


In [16]:
def sql_category_summary(work_category: str) -> list[dict]:
    """Summarize rates within a work category."""
    query = """SELECT unit_norm, COUNT(*) AS item_count,
        MIN(rate) AS min_rate, MAX(rate) AS max_rate,
        ROUND(AVG(rate), 0) AS avg_rate
        FROM boq_items
        WHERE work_category = :category
        GROUP BY unit_norm
        ORDER BY item_count DESC
        LIMIT 10"""
    with engine.connect() as conn:
        rows = conn.execute(text(query), {"category": work_category}).fetchall()
    return [dict(r._mapping) for r in rows]

for cat in ["civil_id", "electrical_elv", "hvac", "fire_fighting", "variation"]:
    print(f"\n=== Category: {cat.upper()} ===")
    rows = sql_category_summary(cat)
    for r in rows:
        mr = 0 if r["min_rate"] is None else r["min_rate"]
        mx = 0 if r["max_rate"] is None else r["max_rate"]
        av = 0 if r["avg_rate"] is None else r["avg_rate"]
        print(f'  unit={r["unit_norm"]:8s} | count={r["item_count"]:3d} | range Rs.{mr:,.0f} - Rs.{mx:,.0f} | avg Rs.{av:,.0f}')



=== Category: CIVIL_ID ===
  unit=sft      | count=385 | range Rs.25 - Rs.10,600 | avg Rs.986
  unit=nos      | count= 98 | range Rs.1,000 - Rs.1,310,500 | avg Rs.63,234
  unit=rft      | count= 57 | range Rs.180 - Rs.34,775 | avg Rs.3,202
  unit=job      | count= 24 | range Rs.3,500 - Rs.1,424,500 | avg Rs.327,104
  unit=ls       | count= 14 | range Rs.7,000 - Rs.1,314,000 | avg Rs.186,929
  unit=points   | count=  6 | range Rs.6,000 - Rs.18,000 | avg Rs.12,867
  unit=cft      | count=  4 | range Rs.750 - Rs.3,235 | avg Rs.2,009
  unit=job.     | count=  3 | range Rs.7,000 - Rs.400,000 | avg Rs.139,000


TypeError: unsupported format string passed to NoneType.__format__

In [13]:
def sql_cross_project(description_fragment: str,
                      unit: str = None) -> list[dict]:
    """
    Show how the same item was priced across projects.
    This is the CORE agent response:
    "In BOQ-01 this was Rs.315, in ID Works Rs.320,
     in 1.Civil.xlsx Rs.310. Consistent pricing."
    """
    query = """
        SELECT 
            description_short,
            rate,
            unit_norm,
            source_file,
            work_category
        FROM boq_items
        WHERE LOWER(description_short) LIKE :pattern
"""
    params = {"pattern": f"{description_fragment.lower()}%"}
    if unit:
        query += " AND unit_norm = :unit"
        params["unit"] = unit
    
    query += " ORDER BY rate ASC"
    
    with engine.connect() as conn:
        rows = conn.execute(text(query), params).fetchall()
    
    return [dict(r._mapping) for r in rows]

# Test — this is exactly what the agent will say
print("=== Cross-project: brick wall 4.5 inch ===")
rows = sql_cross_project("brick work 4.5", unit="sft")
if rows:
    rates = [r["rate"] for r in rows]
    print(f"  Found in {len(rows)} tenders:")
    for r in rows:
        print(f"    Rs. {r['rate']:>6,.0f}/sft  ←  {r['source_file']}")
    print(f"\n  Rate consistency:")
    print(f"    Range: Rs.{min(rates):,.0f} – Rs.{max(rates):,.0f}")
    gap_pct = (max(rates)-min(rates)) / min(rates) * 100
    verdict = "✅ Consistent" if gap_pct < 25 else "⚠️ Review gap"
    print(f"    Gap:   {gap_pct:.1f}%  {verdict}")


=== Cross-project: brick wall 4.5 inch ===
  Found in 1 tenders:
    Rs.    315/sft  ←  BOQ - 01 .xlsx

  Rate consistency:
    Range: Rs.315 – Rs.315
    Gap:   0.0%  ✅ Consistent


In [14]:
def sql_gap_detector(
    item_description: str,
    unit: str,
    market_rate: float
) -> dict:
    """
    Compare historical average vs provided market rate.
    Returns gap analysis the agent will present to user.
    
    market_rate: current rate found via web search (Tavily)
    Returns: gap%, verdict, recommendation
    """
    stats = sql_rate_stats(item_description, unit)
    if not stats or not stats.get("avg_rate"):
        return {"status": "no_history"}
    
    hist_avg = float(stats["avg_rate"])
    gap_pct  = abs(market_rate - hist_avg) / hist_avg * 100
    
    if gap_pct < 15:
        verdict = "✅ CONFIRMED"
        recommendation = (
            f"Historical avg Rs.{hist_avg:,.0f} matches "
            f"market Rs.{market_rate:,.0f} (gap {gap_pct:.1f}%). "
            f"Use historical rate."
        )
    elif gap_pct < 35:
        verdict = "⚠️ MINOR GAP"
        recommendation = (
            f"Historical avg Rs.{hist_avg:,.0f} vs "
            f"market Rs.{market_rate:,.0f} ({gap_pct:.1f}% gap). "
            f"Consider slight adjustment."
        )
    else:
        verdict = "🔴 MAJOR GAP"
        recommendation = (
            f"Historical avg Rs.{hist_avg:,.0f} vs "
            f"market Rs.{market_rate:,.0f} ({gap_pct:.1f}% gap). "
            f"Price inflation likely. Review before submitting."
        )
    
    return {
        "status":          "found",
        "hist_avg":        hist_avg,
        "hist_count":      stats["count"],
        "hist_min":        float(stats["min_rate"]),
        "hist_max":        float(stats["max_rate"]),
        "market_rate":     market_rate,
        "gap_pct":         round(gap_pct, 1),
        "verdict":         verdict,
        "recommendation":  recommendation,
        "sources":         stats["sources"]
    }

# Test gap detector
print("=== Gap Detector Tests ===\n")

# Test 1: Confirmed (market near historical)
r1 = sql_gap_detector("brick work 4.5", "sft", 
                       market_rate=330.0)
print(f"Brick work — market Rs.330:")
print(f"  {r1['verdict']}")
print(f"  {r1['recommendation']}\n")

# Test 2: Major gap (market much higher)
r2 = sql_gap_detector("gypsum", "sft", 
                       market_rate=850.0)
print(f"Gypsum ceiling — market Rs.850:")
print(f"  {r2['verdict']}")
print(f"  {r2['recommendation']}\n")

# Test 3: No history
r3 = sql_gap_detector("solar panel installation", 
                       "nos", market_rate=50000.0)
print(f"Solar panel — no history:")
print(f"  Status: {r3['status']}")


=== Gap Detector Tests ===

Brick work — market Rs.330:
  ✅ CONFIRMED
  Historical avg Rs.315 matches market Rs.330 (gap 4.8%). Use historical rate.

Gypsum ceiling — market Rs.850:
  🔴 MAJOR GAP
  Historical avg Rs.423 vs market Rs.850 (100.9% gap). Price inflation likely. Review before submitting.

Solar panel — no history:
  Status: no_history


## Step 5: Save PostgreSQL Module

Save all SQL functions to src/rateiq/postgres_store.py
so the LangGraph agent can import them.

In [15]:
MODULE_PATH = Path("../src/rateiq/postgres_store.py")

module_code = '''"""
RateIQ PostgreSQL Rate History Module
Structured SQL queries for rate statistics and gap detection.
Import this in the LangGraph agent alongside BOQSearcher.
"""
import json, re
import numpy as np
import pandas as pd
from sqlalchemy import create_engine, text
from dotenv import load_dotenv
import os

load_dotenv()

POSTGRES_URL = os.getenv(
    "POSTGRES_URL",
    "postgresql://rateiq:rateiq123@localhost:5432/rateiq"
)

_engine = None

def get_engine():
    """Lazy singleton engine."""
    global _engine
    if _engine is None:
        _engine = create_engine(POSTGRES_URL)
    return _engine


def sql_rate_lookup(description_fragment: str,
                    unit: str = None,
                    limit: int = 5) -> list:
    """Find items by description keyword. Returns list of dicts."""
    q = """SELECT description_short, rate, unit_norm,
                  source_file, rate_per_unit_label, work_category
           FROM boq_items
           WHERE LOWER(description_short) LIKE :pattern"""
    p = {"pattern": f"{description_fragment.lower()}%"}
    if unit:
        q += " AND unit_norm = :unit"; p["unit"] = unit
    q += " ORDER BY rate ASC LIMIT :limit"; p["limit"] = limit
    with get_engine().connect() as c:
        rows = c.execute(text(q), p).fetchall()
    return [dict(r._mapping) for r in rows]


def sql_rate_stats(description_fragment: str,
                   unit: str = None) -> dict:
    """Statistical summary for an item type."""
    q = """SELECT COUNT(*) AS count,
                  MIN(rate) AS min_rate, MAX(rate) AS max_rate,
                  ROUND(AVG(rate),0) AS avg_rate,
                  ROUND(STDDEV(rate),0) AS std_rate,
                  ARRAY_AGG(DISTINCT source_file) AS sources
           FROM boq_items
           WHERE LOWER(description_short) LIKE :pattern"""
    p = {"pattern": f"{description_fragment.lower()}%"}
    if unit:
        q += " AND unit_norm = :unit"; p["unit"] = unit
    with get_engine().connect() as c:
        row = c.execute(text(q), p).fetchone()
    if not row or row.count == 0:
        return {}
    return dict(row._mapping)


def sql_cross_project(description_fragment: str,
                      unit: str = None) -> list:
    """Cross-project rate comparison."""
    q = """SELECT description_short, rate, unit_norm,
                  source_file, work_category
           FROM boq_items
           WHERE LOWER(description_short) LIKE :pattern"""
    p = {"pattern": f"{description_fragment.lower()}%"}
    if unit:
        q += " AND unit_norm = :unit"; p["unit"] = unit
    q += " ORDER BY rate ASC"
    with get_engine().connect() as c:
        rows = c.execute(text(q), p).fetchall()
    return [dict(r._mapping) for r in rows]


def sql_gap_detector(description_fragment: str,
                     unit: str,
                     market_rate: float) -> dict:
    """Compare historical avg vs market rate. Returns gap analysis."""
    stats = sql_rate_stats(description_fragment, unit)
    if not stats or not stats.get("avg_rate"):
        return {"status": "no_history"}
    hist_avg = float(stats["avg_rate"])
    gap_pct  = abs(market_rate - hist_avg) / hist_avg * 100
    if gap_pct < 15:
        verdict = "CONFIRMED"
        rec = (f"Historical avg Rs.{hist_avg:,.0f} matches "
               f"market Rs.{market_rate:,.0f} (gap {gap_pct:.1f}%).")
    elif gap_pct < 35:
        verdict = "MINOR_GAP"
        rec = (f"Historical Rs.{hist_avg:,.0f} vs "
               f"market Rs.{market_rate:,.0f} ({gap_pct:.1f}% gap).")
    else:
        verdict = "MAJOR_GAP"
        rec = (f"Historical Rs.{hist_avg:,.0f} vs "
               f"market Rs.{market_rate:,.0f} ({gap_pct:.1f}% gap). "
               f"Price inflation likely.")
    return {
        "status": "found", "hist_avg": hist_avg,
        "hist_count": stats["count"],
        "hist_min": float(stats["min_rate"]),
        "hist_max": float(stats["max_rate"]),
        "market_rate": market_rate,
        "gap_pct": round(gap_pct, 1),
        "verdict": verdict, "recommendation": rec,
        "sources": stats.get("sources", [])
    }
'''

MODULE_PATH.write_text(module_code)
print(f"✓ Saved: {MODULE_PATH}")

# Test import
import sys
sys.path.insert(0, str(Path("../src").resolve()))

for k in list(sys.modules.keys()):
    if "rateiq" in k:
        del sys.modules[k]

from rateiq.postgres_store import (
    sql_rate_lookup, sql_rate_stats, 
    sql_cross_project, sql_gap_detector
)

test = sql_rate_lookup("brick work", unit="sft", limit=2)
print(f"Import test: {test[0]['description_short'][:40]}")
print("✓ postgres_store module working")

# Update __init__.py
init_path = Path("../src/rateiq/__init__.py")
init_path.write_text(
    "from .hybrid_search import BOQSearcher\n"
    "from .postgres_store import (\n"
    "    sql_rate_lookup, sql_rate_stats,\n"
    "    sql_cross_project, sql_gap_detector\n"
    "\n"
)
print("✓ __init__.py updated")


✓ Saved: ../src/rateiq/postgres_store.py
Import test: Brick Work 4.5" Thick
✓ postgres_store module working
✓ __init__.py updated


## Notebook 05 Complete

| Table            | Rows  | Purpose                        |
|------------------|-------|--------------------------------|
| boq_items        | 1429  | Raw line items, full metadata  |
| boq_rate_stats   | N     | Pre-aggregated stats per type  |

SQL Functions saved to src/rateiq/postgres_store.py:
- sql_rate_lookup()    → find items by keyword
- sql_rate_stats()     → avg/min/max for item type
- sql_cross_project()  → rates across all projects
- sql_gap_detector()   → compare historical vs market

## Agent Will Use These Together:

Step 1: BOQSearcher.smart_search(description)
        → semantic match, top-5 candidates

Step 2: sql_cross_project(keyword, unit)
        → "In 3 past tenders: Rs.310, Rs.315, Rs.320"

Step 3: sql_gap_detector(keyword, unit, market_rate)
        → "✅ CONFIRMED: historical matches market"

Step 4: Agent assembles final answer with source citations

## Next: Notebook 06 — LangGraph Agent
The agent that calls all these tools and fills the BOQ.